instrumentation/opentelemetry-instrumentation-flask/tests/test_programmatic.py

```diff
@@ -243,3 +244,87 @@ def test_custom_span_name(self):
         span_list = self.memory_exporter.get_finished_spans()
         self.assertEqual(len(span_list), 1)
         self.assertEqual(span_list[0].name, "instrument-without-app")
+
+class TestProgrammaticHooks(
+    InstrumentationTest, TestBase, WsgiTestBase
+):
+    def setUp(self):
+        super().setUp()
+
+        hook_headers = (
+            "hook_attr",
+            "hello otel",
+        )
+
+        def request_hook(span, environ):
+            span.update_name("name from hook")
+
+        def response_hook(span, environ, response_headers):
+            span.set_attribute("hook_attr", "hello world")
+            response_headers.append(hook_headers)
+        
+        self.app = Flask(__name__)
+
+        FlaskInstrumentor().instrument_app(
+            self.app, otel_request_hook=request_hook, otel_response_hook=response_hook
+        )
+
+        self._common_initialization()
+    
+    def tearDown(self):
+        super().tearDown()
+        with self.disable_logging():
+            FlaskInstrumentor().uninstrument_app(self.app)
+    
+    def test_hooks(self):
+        expected_attrs = expected_attributes(
+            {"http.target": "/hello/123", "http.route": "/hello/<int:helloid>", "hook_attr":"hello world"}
+        )
+
+        self.client.get("/hello/123")
```

Comment:
I would also test adding a header to the response and making sure it was seen here.

Reply:
If you don't mind, could you elaborate on how to do that?

Reply:
You'd need to update the post response hook and inject headers into the response there. Look at Flask or WSGI docs (which ever object is passed to the response hook). Then you'd capture the result from `client.get()` call here and verify the result has the same headers you set in the hook earlier.